src/patterns/pattern/mod.rs

```diff
@@ -27,19 +27,58 @@ use board::White;
 
 mod test;
 
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum PointPattern {
+    Black,
+    NotBlack,
+    White,
+    NotWhite,
+    All,
+    Empty,
+    OffBoard
+}
+
+impl PointPattern {
+
+    pub fn from_char(c: &char) -> PointPattern {
+        match *c {
+            'X' => PointPattern::Black,
+            'O' => PointPattern::White,
+            '?' => PointPattern::All,
+            'x' => PointPattern::NotBlack,
+            'o' => PointPattern::NotWhite,
+            '.' => PointPattern::Empty,
+            ' ' => PointPattern::OffBoard,
+            _   => panic!("Can't convert {:?} to PointPattern", c)
+        }
+    }
+
+}
+
 #[derive(Clone, Debug, Eq, PartialEq)]
 pub struct Pattern {
-    vec: Vec<Vec<char>>
+    vec: Vec<PointPattern>
 }
 
 impl Pattern {
 
     pub fn new(vec: Vec<Vec<char>>) -> Pattern {
+        let mut v = vec!();
+        // Can be done with flat_map, I think.
+        for sv in vec.iter() {
+            for c in sv.iter() {
+                v.push(PointPattern::from_char(c));
+            }
+        }
+        Pattern { vec: v }
+    }
+
+    fn raw(vec: Vec<PointPattern>) -> Pattern {
         Pattern { vec: vec }
     }
 
     pub fn matches(&self, board: &Board, coord: &Coord) -> bool {
-        coord.neighbours8_unchecked()
+        board.neighbours8_unchecked(*coord)
```

Comment:
This was the first optimisation. Recalculating the neighbours of each `Coord` for each pattern is of course a waste of time.